### PR TITLE
EOSC Accounting

### PR DESCRIPTION
**Pull request by enolfc**
# Summary

Implement changes needed to make Notebooks report accounting as needed by EOSC Lot 1/3
Modernise the code as well

Changes to the main accounting script:
- it changes the configuration and splits the configuration into a `[default]` section and keeps there most of the stuff that was on `[prometheus]` which was a mix of everything
- the DB is updated with a new field (`flavor`) which is populated from the `egi.eu/flavor` annotation
- APEL records are not generated by default, needs to be explicitly enabled in the config
- default VO is removed, not clear that it was used at all

In the EOSC accounting:
- it does not query prometheus directly anymore but does queries to the existing database for VM that have
- it can use a `from-date` and `to-date` for reporting, by default it will report the previous day
- uses client credentials grant instead of refresh 

Other changes
- it moves to `pyproject.toml`, we may think on additional tooling for managing the project in the future

---

**Related issue :**

### CODE REVIEW

File: egi_notebooks_accounting/eosc.py [hunk 1]
```diff
@@ -103,6 +106,80 @@ def update_pod_metric(pod, metrics, flavor_config):
...
+        report_day = date.today() - timedelta(days=1)
+        to_date = datetime(report_day.year, report_day.month, report_day.day, 23, 59)
+    utc = pytz.UTC
+    from_date = from_date.replace(tzinfo=utc)
+    to_date = to_date.replace(tzinfo=utc)
+    return from_date, to_date
+
+
+def generate_day_metrics(
+    period_start, period_end, accounting_url, token, flavor_config, dry_run
```
Comment (andrea-manzi) [inline egi_notebooks_accounting/eosc.py]: need to add  _installation_ and _timestamp_file_ as argument

Comment (andrea-manzi) [reply egi_notebooks_accounting/eosc.py]: pushed changes

**Comment by andrea-manzi** [comment]
i'm testing manullay for now, but i see some strange aggregated values being published to Accounting:

`DEBUG:root:Sending metric {'metric_definition_id': '668bdd5988e1d617b217ecb9', 'time_period_start': '2024-07-25T00:00:00Z', 'time_period_end': '2024-07-25T23:59:00Z', 'user': '3097fd08-1b55-462a-b4ee-95613998655f@eosc-federation.eu', 'group': 'urn:geant:eosc-federation.eu:testing:group:eosc#testing.eosc-federation.eu', 'value': 57002.0} to accounting
DEBUG:root:Pushing to accounting - 668bdf9c744c3c7ae7531916
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): api.acc.staging.eosc.grnet.gr:443
DEBUG:urllib3.connectionpool:https://api.acc.staging.eosc.grnet.gr:443 "POST /accounting-system/installations/668bdf9c744c3c7ae7531916/metrics HTTP/1.1" 201 190
DEBUG:root:Sending metric {'metric_definition_id': '668bdd5988e1d617b217ecb9', 'time_period_start': '2024-07-25T00:00:00Z', 'time_period_end': '2024-07-25T23:59:00Z', 'user': '6d52258d-4af3-4a0b-ac9f-dd3cb9d0a4dd@eosc-federation.eu', 'group': 'pp-5e869742-24f1-4ae0-a21c-13a64cfe7115', 'value': 7751.0} to accounting
DEBUG:root:Pushing to accounting - 668bdf9c744c3c7ae7531916
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): api.acc.staging.eosc.grnet.gr:443
DEBUG:urllib3.connectionpool:https://api.acc.staging.eosc.grnet.gr:443 "POST /accounting-system/installations/668bdf9c744c3c7ae7531916/metrics HTTP/1.1" 201 189
DEBUG:root:Sending metric {'metric_definition_id': '668bdd5988e1d617b217ecb9', 'time_period_start': '2024-07-25T00:00:00Z', 'time_period_end': '2024-07-25T23:59:00Z', 'user': 'e37e2773-c993-4cee-aa7e-1944089d9716@eosc-federation.eu', 'group': 'pp-d71db0cf-3d77-4f97-976d-8fa5a51fe3ff', 'value': 96.0} to accounting` 

i have generated the  DB from prometheus on fullhub instance using the pods py script

**Comment by andrea-manzi** [comment]
> i'm testing manullay for now, but i see some strange aggregated values being published to Accounting:
> 
> `DEBUG:root:Sending metric {'metric_definition_id': '668bdd5988e1d617b217ecb9', 'time_period_start': '2024-07-25T00:00:00Z', 'time_period_end': '2024-07-25T23:59:00Z', 'user': '3097fd08-1b55-462a-b4ee-95613998655f@eosc-federation.eu', 'group': 'urn:geant:eosc-federation.eu:testing:group:eosc#testing.eosc-federation.eu', 'value': 57002.0} to accounting DEBUG:root:Pushing to accounting - 668bdf9c744c3c7ae7531916 DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): api.acc.staging.eosc.grnet.gr:443 DEBUG:urllib3.connectionpool:https://api.acc.staging.eosc.grnet.gr:443 "POST /accounting-system/installations/668bdf9c744c3c7ae7531916/metrics HTTP/1.1" 201 190 DEBUG:root:Sending metric {'metric_definition_id': '668bdd5988e1d617b217ecb9', 'time_period_start': '2024-07-25T00:00:00Z', 'time_period_end': '2024-07-25T23:59:00Z', 'user': '6d52258d-4af3-4a0b-ac9f-dd3cb9d0a4dd@eosc-federation.eu', 'group': 'pp-5e869742-24f1-4ae0-a21c-13a64cfe7115', 'value': 7751.0} to accounting DEBUG:root:Pushing to accounting - 668bdf9c744c3c7ae7531916 DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): api.acc.staging.eosc.grnet.gr:443 DEBUG:urllib3.connectionpool:https://api.acc.staging.eosc.grnet.gr:443 "POST /accounting-system/installations/668bdf9c744c3c7ae7531916/metrics HTTP/1.1" 201 189 DEBUG:root:Sending metric {'metric_definition_id': '668bdd5988e1d617b217ecb9', 'time_period_start': '2024-07-25T00:00:00Z', 'time_period_end': '2024-07-25T23:59:00Z', 'user': 'e37e2773-c993-4cee-aa7e-1944089d9716@eosc-federation.eu', 'group': 'pp-d71db0cf-3d77-4f97-976d-8fa5a51fe3ff', 'value': 96.0} to accounting`
> 
> i have generated the DB from prometheus on fullhub instance using the pods py script

could be that this is related to this query https://github.com/EGI-Federation/egi-notebooks-accounting/blob/master/egi_notebooks_accounting/pods.py#L46 which is returning seconds? while we would like to report hours?

**Comment by andrea-manzi** [comment]
> > i'm testing manullay for now, but i see some strange aggregated values being published to Accounting:
> > `DEBUG:root:Sending metric {'metric_definition_id': '668bdd5988e1d617b217ecb9', 'time_period_start': '2024-07-25T00:00:00Z', 'time_period_end': '2024-07-25T23:59:00Z', 'user': '3097fd08-1b55-462a-b4ee-95613998655f@eosc-federation.eu', 'group': 'urn:geant:eosc-federation.eu:testing:group:eosc#testing.eosc-federation.eu', 'value': 57002.0} to accounting DEBUG:root:Pushing to accounting - 668bdf9c744c3c7ae7531916 DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): api.acc.staging.eosc.grnet.gr:443 DEBUG:urllib3.connectionpool:https://api.acc.staging.eosc.grnet.gr:443 "POST /accounting-system/installations/668bdf9c744c3c7ae7531916/metrics HTTP/1.1" 201 190 DEBUG:root:Sending metric {'metric_definition_id': '668bdd5988e1d617b217ecb9', 'time_period_start': '2024-07-25T00:00:00Z', 'time_period_end': '2024-07-25T23:59:00Z', 'user': '6d52258d-4af3-4a0b-ac9f-dd3cb9d0a4dd@eosc-federation.eu', 'group': 'pp-5e869742-24f1-4ae0-a21c-13a64cfe7115', 'value': 7751.0} to accounting DEBUG:root:Pushing to accounting - 668bdf9c744c3c7ae7531916 DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): api.acc.staging.eosc.grnet.gr:443 DEBUG:urllib3.connectionpool:https://api.acc.staging.eosc.grnet.gr:443 "POST /accounting-system/installations/668bdf9c744c3c7ae7531916/metrics HTTP/1.1" 201 189 DEBUG:root:Sending metric {'metric_definition_id': '668bdd5988e1d617b217ecb9', 'time_period_start': '2024-07-25T00:00:00Z', 'time_period_end': '2024-07-25T23:59:00Z', 'user': 'e37e2773-c993-4cee-aa7e-1944089d9716@eosc-federation.eu', 'group': 'pp-d71db0cf-3d77-4f97-976d-8fa5a51fe3ff', 'value': 96.0} to accounting`
> > i have generated the DB from prometheus on fullhub instance using the pods py script
> 
> could be that this is related to this query https://github.com/EGI-Federation/egi-notebooks-accounting/blob/master/egi_notebooks_accounting/pods.py#L46 which is returning seconds? while we would like to report hours?

mm no that value is coming from the pod.wall

**Comment by enolfc** [comment]
> 
> mm no that value is coming from the pod.wall

that's also in seconds, so we need to convert to hours

**Comment by andrea-manzi** [comment]
> > mm no that value is coming from the pod.wall
> 
> that's also in seconds, so we need to convert to hours

yes now numbers seems correct!

```
DEBUG:urllib3.connectionpool:https://api.acc.staging.eosc.grnet.gr:443 "POST /accounting-system/installations/668bdf9c744c3c7ae7531916/metrics HTTP/1.1" 201 200
DEBUG:root:Sending metric {'metric_definition_id': '668bdd5988e1d617b217ecb9', 'time_period_start': '2024-07-25T00:00:00Z', 'time_period_end': '2024-07-25T23:59:00Z', 'user': '6d52258d-4af3-4a0b-ac9f-dd3cb9d0a4dd@eosc-federation.eu', 'group': 'pp-5e869742-24f1-4ae0-a21c-13a64cfe7115', 'value': 2.1530555555555555} to accounting
DEBUG:root:Pushing to accounting - 668bdf9c744c3c7ae7531916
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): api.acc.staging.eosc.grnet.gr:443
DEBUG:urllib3.connectionpool:https://api.acc.staging.eosc.grnet.gr:443 "POST /accounting-system/installations/668bdf9c744c3c7ae7531916/metrics HTTP/1.1" 201 201
DEBUG:root:Sending metric {'metric_definition_id': '668bdd5988e1d617b217ecb9', 'time_period_start': '2024-07-25T00:00:00Z', 'time_period_end': '2024-07-25T23:59:00Z', 'user': 'e37e2773-c993-4cee-aa7e-1944089d9716@eosc-federation.eu', 'group': 'pp-d71db0cf-3d77-4f97-976d-8fa5a51fe3ff', 'value': 0.02666666666666667} to accounting
```
@valtri do you think you can test it on the testing JHub instance ?

**Review by andrea-manzi — APPROVED** [review]
Tested successfully by CESNET

**Pull request merged**